src/internal/pjs/server_internal_test.go

```diff
@@ -0,0 +1,480 @@
+		if status.Code(err) == codes.Canceled {
+			err = nil
+		}
+		return err
+	})
+	var ret *pjs.JobInfo_Success
+	eg.Go(func() error {
+		jobInfo, err := await(ctx, c, from)
+		if err != nil {
+			return err
+		}
+		ret = jobInfo.GetSuccess()
+		cf() // success, cancel the other gorountine
+		return nil
+	})
+	err = eg.Wait()
+	require.NoError(t, err)
+	return ret, nil
+}
+
+func processQueue(pqc pjs.API_ProcessQueueClient, programHash []byte, fn func(resp *pjs.ProcessQueueResponse) error) error {
+	if err := pqc.Send(&pjs.ProcessQueueRequest{
+		Queue: &pjs.Queue{Id: programHash},
+	}); err != nil {
+		return err
+	}
+	for {
+		msg, err := pqc.Recv()
+		if err != nil {
+			return err
+		}
+		err = fn(msg)
+		if err != nil {
+			if sendErr := pqc.Send(&pjs.ProcessQueueRequest{
+				Result: &pjs.ProcessQueueRequest_Failed{
+					Failed: true,
+				},
+			}); sendErr != nil {
+				return sendErr
+			}
+		} else {
+			// todo(muayng): if send returns an error, the ctx on the server side will be cancelled. Job cannot transit to
+			// Done state. Await will spin forever.
+			if err := pqc.Send(&pjs.ProcessQueueRequest{
+				Result: &pjs.ProcessQueueRequest_Success_{
+					Success: &pjs.ProcessQueueRequest_Success{
+						Output: msg.Input,
+					},
+				},
+			}); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// await blocks until a Job enters the DONE state
+func await(ctx context.Context, s pjs.APIClient, jid int64) (*pjs.JobInfo, error) {
+	for {
+		res, err := s.InspectJob(ctx, &pjs.InspectJobRequest{
+			Job: &pjs.Job{Id: jid},
+		})
+		if err != nil {
+			return nil, err
+		}
+		if res.Details.JobInfo.State == pjs.JobState_DONE {
+			return res.Details.JobInfo, nil
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+func setupTest(t testing.TB) (pjs.APIClient, storage.FilesetClient) {
+	ctx := pctx.TestContext(t)
+	db := dockertestenv.NewTestDB(t)
+	migrationEnv := migrations.Env{EtcdClient: testetcd.NewEnv(ctx, t).EtcdClient}
+	require.NoError(t, migrations.ApplyMigrations(ctx, db, migrationEnv, clusterstate.DesiredClusterState), "should be able to set up tables")
+	return NewTestClient(t, db), storagesrv.NewTestFilesetClient(t, db)
+}
+
+func createFileSet(t *testing.T, fc storage.FilesetClient, files map[string][]byte) string {
+	ctx := pctx.TestContext(t)
+	ctx, cf := context.WithCancel(ctx)
+	defer cf()
+	createFileClient, err := fc.CreateFileset(ctx)
+	require.NoError(t, err)
+	for path, data := range files {
+		require.NoError(t, createFileClient.Send(&storage.CreateFilesetRequest{
+			Modification: &storage.CreateFilesetRequest_AppendFile{
+				AppendFile: &storage.AppendFile{
+					Path: path,
+					Data: &wrapperspb.BytesValue{
+						Value: data,
+					},
+				},
+			},
+		}))
+	}
+	resp, err := createFileClient.CloseAndRecv()
+	require.NoError(t, err)
+	return resp.FilesetId
+}
+
+type createJobOpts func(req *pjs.CreateJobRequest)
+
+func createJob(t *testing.T, ctx context.Context, c pjs.APIClient, fc storage.FilesetClient,
```

Comment:
FWIW, I don't think this helper is worth the cognitive load of being a helper function.  Some sort of constructor for a CreateJobRequest with dummy inputs (`func programAndInputs(programText string, inputs map[string][]byte) *pjs.CreateJobRequest) would be easier.

The reason is that with the mutating opts, this isn't generic enough to work every time, and it would be easier to simply mutate the request in the test, so there is no action at a distance.  You have to visit this function and read it in order to call it, I'm not sure it's worth the trip for something so simple.

Reply:
I agree. I think its worth just allowing you to pass in a request and providing another function for a default request.